qiskit_experiments/calibration/calibrations.py

```diff
@@ -0,0 +1,972 @@
+                if re.compile(self.__channel_pattern__).match(ch.index.name) is None:
+                    raise CalibrationError(
+                        f"Parameterized channel must correspond to {self.__channel_pattern__}"
+                    )
+
+        # Check that subroutines are present.
+        for block in schedule.blocks:
+            if isinstance(block, Call):
+                if isinstance(block.subroutine, Schedule):
+                    raise CalibrationError(
+                        "Calling a Schedule is forbidden, call ScheduleBlock instead."
+                    )
+
+                if (block.subroutine.name, qubits) not in self._schedules:
+                    raise CalibrationError(
+                        f"Cannot register schedule block {schedule.name} with unregistered "
+                        f"subroutine {block.subroutine.name}."
+                    )
+
+        # Clean the parameter to schedule mapping. This is needed if we overwrite a schedule.
+        self._clean_parameter_map(schedule.name, qubits)
+
+        # Add the schedule.
+        self._schedules[ScheduleKey(schedule.name, qubits)] = schedule
+
+        # Register parameters that are not indices.
+        # Do not register parameters that are in call instructions.
+        params_to_register = set()
+        for inst in self._exclude_calls(schedule, []):
+            for param in inst.parameters:
+                if param not in param_indices:
+                    params_to_register.add(param)
+
+        if len(params_to_register) != len(set(param.name for param in params_to_register)):
+            raise CalibrationError(f"Parameter names in {schedule.name} must be unique.")
+
+        for param in params_to_register:
+            self._register_parameter(param, qubits, schedule)
+
+    def _exclude_calls(
+        self, schedule: ScheduleBlock, instructions: List[Instruction]
+    ) -> List[Instruction]:
+        """Return the non-Call instructions.
+
+        Recursive function to get all non-Call instructions. This will flatten all blocks
+        in a :class:`ScheduleBlock` and return the instructions of the ScheduleBlock leaving
+        out any Call instructions.
+
+        Args:
+            schedule: A :class:`ScheduleBlock` from which to extract the instructions.
+            instructions: The list of instructions that is recursively populated.
+
+        Returns:
+            The list of instructions to which all non-Call instructions have been added.
+        """
+        for block in schedule.blocks:
+            if isinstance(block, ScheduleBlock):
+                instructions = self._exclude_calls(block, instructions)
+            else:
+                if not isinstance(block, Call):
+                    instructions.append(block)
+
+        return instructions
+
+    def get_template(
+        self, schedule_name: str, qubits: Optional[Tuple[int, ...]] = None
+    ) -> ScheduleBlock:
+        """Get a template schedule.
```

Comment:
Can you elaborate what is a _template_ schedule? It seems we can get the same object by calling `get_schedule` without `assign_params` because it sounds like we don't assign anything. But this is not true. Once we read the code, we'll realize we should actively assign parameter objects to all keys the schedule has. On the other hand, this will return unassigned schedule.

Reply:
I'll clarify the different mechanisms of get_schedule and get_template in the doc strings.

Reply:
See here: bab9b9a